jax/_src/pallas/mosaic_gpu/lowering.py

```diff
@@ -2152,3 +2159,144 @@ def _ensure_idx_fa(x):
+    )
+  device_id = _device_id_to_logical(ctx, device_id, device_id_type)
+
+  i64_ty = ir.IntegerType.get_signless(64)
+  i32_ty = ir.IntegerType.get_signless(32)
+  sem_ptr = mgpu.utils.memref_ptr(sem)
+  val = _ir_constant(value, i32_ty)
+  mgpu_utils.warpgroup_barrier()
+  with mgpu.single_thread(per_block=False):
+    mgpu_utils.system_membar()
+    llvm.CallOp(None, [sem_ptr, val, mgpu_core.device_id_to_llvm(device_id)], op_bundle_operands=[], op_bundle_sizes=[], op_bundle_tags=[], callee="nvshmem_int32_atomic_add")
+  return ()
+
+
+@register_lowering_rule(primitives.semaphore_wait_p)
+def _semaphore_wait_lowering_rule(ctx: LoweringRuleContext, *args, args_tree):
+  sem_aval, _, _ = tree_util.tree_unflatten(args_tree, ctx.avals_in)
+  sem, transforms, value = tree_util.tree_unflatten(args_tree, args)
+  sem, _ = _handle_indexing(sem, transforms)
+
+  sem_ptr = mgpu.utils.memref_ptr(sem)
+  i64_ty = ir.IntegerType.get_signless(64)
+  i32_ty = ir.IntegerType.get_signless(32)
+  NVSHMEM_CMP_EQ = mgpu.utils.c(0, i32_ty)
+  val = _ir_constant(value, i32_ty)
+  with mgpu.single_thread(per_block=False):
+    # Wait semaphone
+    llvm.CallOp(None, [sem_ptr, NVSHMEM_CMP_EQ, val], op_bundle_operands=[], op_bundle_sizes=[], op_bundle_tags=[], callee="nvshmem_int32_wait_until")
+    # Reset semaphone
+    minus_val = arith_dialect.muli(_ir_constant(-1, i32_ty), val)
+    addi_enum_val = 1
+    memref_dialect.atomic_rmw(addi_enum_val, minus_val, sem, [])
+  mgpu_utils.warpgroup_barrier()
+  return ()
+
+
+@register_lowering_rule(primitives.dma_start_p)
+def _dma_start_lowering_rule(ctx: LoweringRuleContext, *args, tree,
+                             device_id_type: primitives.DeviceIdType):
+  (
+      src_ref,
+      src_transforms,
+      dst_ref,
+      dst_transforms,
+      sem,
+      sem_transforms,
+      src_sem,
+      src_sem_transforms,
+      device_id,
+  ) = tree_util.tree_unflatten(tree, args)
+  (src_ref_aval, _, dst_ref_aval, _, sem_aval, _, src_sem_aval, _, _) = (
+      tree_util.tree_unflatten(tree, ctx.avals_in)
+  )
+  if src_ref_aval.dtype == jnp.bool_:
+    raise NotImplementedError("DMAs with bool dtypes are not supported.")
+
+  src_ref, _ = _handle_indexing(src_ref, src_transforms)
+
+  if src_sem is not None:
+    src_sem, _ = _handle_indexing(src_sem, src_sem_transforms)
+  dst_ref, _ = _handle_indexing(dst_ref, dst_transforms)
+  sem, _ = _handle_indexing(sem, sem_transforms)
+
+
+  if device_id is None:
+    raise ValueError(
+      "'async_remote_copy' requires 'device_id' argument to be set on GPUs."
+    )
+  device_id = _device_id_to_logical(ctx, device_id, device_id_type)
+
+  ref_ty = ir.MemRefType(src_ref.type)
+  i64_ty = ir.IntegerType.get_signless(64)
+  i32_ty = ir.IntegerType.get_signless(32)
+  element_bytes = mgpu.utils.c(
+   mgpu.utils.bytewidth(ref_ty.element_type), i64_ty
+  )
+  mgpu.utils.extern_elementwise_func_apply("nvshmem_putmem", None,
```

Comment:
I thought the primary benefit of nvshmem is that we'll be able to have consistent address spaces across devices and will be able to just copy data between them using regular loads and stores, or even better using `cp.async.bulk`. Is that not the case? The approach of emitting a loop that does lots of small transfers seems to have a fairly high overhead to me